File: apple/tvos/tvOSSampleApp_ObjC/tvOSSampleApp_ObjC/CollectionCell.m
//
//  CollectionCell.m
//  tvOSSampleApp_ObjC
//
//  Created by Patrick McConnell on 2/18/16.
//  Copyright © 2016 Patrick McConnell. All rights reserved.
//

#import "CollectionCell.h"
#import "tvOSSampleApp_ObjC-Swift.h"

@implementation CollectionCell

- (void) configureWithCollection: (OddMediaObjectCollection*) collection {
  [self.titleLabel setText:collection.title];
  
  [collection thumbnail:^(UIImage* image) {
    if (image != nil) {
      dispatch_async(dispatch_get_main_queue(), ^{
        [self.thumbnailImageView setImage:image];
      });
    }
  }];
}

- (void) becomeFocusedUsingAnimationCoordinator: (UIFocusAnimationCoordinator*) coordinator {
  [coordinator addCoordinatedAnimations:^{
    [self setTransform: CGAffineTransformMakeScale(1.1, 1.1) ];
    [self.layer setShadowColor: [[UIColor blackColor] CGColor] ];
    [self.layer setShadowOffset: CGSizeMake(10, 10) ];
    [self.layer setShadowOpacity: 0.2 ];
    [self.layer setShadowRadius: 5 ];
  } completion:^{
    
  }];
}

- (void) resignFocusedUsingAnimationCoordinator: (UIFocusAnimationCoordinator*) coordinator {
  [coordinator addCoordinatedAnimations:^{
    [self setTransform: CGAffineTransformIdentity ];
    [self.layer setShadowColor: nil ];
    [self.layer setShadowOffset: CGSizeZero ];
  } completion:^{
    
  }];
}


-(void)didUpdateFocusInContext:(UIFocusUpdateContext *)context withAnimationCoordinator:(UIFocusAnimationCoordinator *)coordinator {
  [super didUpdateFocusInContext:context withAnimationCoordinator:coordinator];
  
  if ([context nextFocusedView] == nil) { return; }
  
  CollectionCell* nextFocusedView = (CollectionCell*)[context nextFocusedView];
  
  if (nextFocusedView == self) {
    [self becomeFocusedUsingAnimationCoordinator:coordinator];
    [(UIView*)self addParallaxMotionEffects:0.5 panValue:5 ];
  } else {
    [self resignFocusedUsingAnimationCoordinator:coordinator];
    [self setMotionEffects: @[] ];
  }
}




@end

File: apple/tvos/tvOSSampleApp/tvOSSampleApp/CollectionCell.swift
//
//  CollectionCell.swift
//  tvOSSampleApp
//
//  Created by Patrick McConnell on 1/28/16.
//  Copyright © 2016 Odd Networks. All rights reserved.
//

import UIKit
import OddSDKtvOS

class CollectionCell: UICollectionViewCell {
  @IBOutlet weak var thumbnailImageView: UIImageView!
  @IBOutlet weak var titleLabel: UILabel!
  
  func configureWithCollection(_ collection: OddMediaObjectCollection) {
    
    self.titleLabel?.text = collection.title
    
    collection.thumbnail { (image) -> Void in
      if let thumbnail = image {
        DispatchQueue.main.async(execute: { () -> Void in
          self.thumbnailImageView?.image = thumbnail
        })
      }
    }
  }
  
  func becomeFocusedUsingAnimationCoordinator(_ coordinator: UIFocusAnimationCoordinator) {
    coordinator.addCoordinatedAnimations({ () -> Void in
      self.transform = CGAffineTransform(scaleX: 1.1, y: 1.1)
      self.layer.shadowColor = UIColor.black.cgColor
      self.layer.shadowOffset = CGSize(width: 10, height: 10)
      self.layer.shadowOpacity = 0.2
      self.layer.shadowRadius = 5
      }) { () -> Void in }
  }
  
  func resignFocusUsingAnimationCoordinator(_ coordinator: UIFocusAnimationCoordinator) {
    coordinator.addCoordinatedAnimations({ () -> Void in
      self.transform = CGAffineTransform.identity
      self.layer.shadowColor = nil
      self.layer.shadowOffset = CGSize.zero
      }) { () -> Void in }
  }
  
  override func didUpdateFocus(in context: UIFocusUpdateContext, with coordinator: UIFocusAnimationCoordinator) {
    super.didUpdateFocus(in: context, with: coordinator)
    
    guard let nextFocusedView = context.nextFocusedView else { return }
    
    if nextFocusedView == self {
      self.becomeFocusedUsingAnimationCoordinator(coordinator)
      self.addParallaxMotionEffects()
    } else {
      self.resignFocusUsingAnimationCoordinator(coordinator)
      self.motionEffects = []
    }
  }

    
}
